Fix authorization rules to the Assistant and UserPreferences service

jakule opened the pull request:
Closes https://github.com/gravitational/teleport-private/issues/797

jakule committed:
"Add authorization rules to the Assistant and UserPreferences service"

This commit introduces authorization rules into the Assistant service to restrict operations based on the authenticated user's role permissions.

Now each method in the Assistant service checks if the authenticated user has necessary permissions to perform the requested operation. The permissions are checked via defined RBAC rules.

A user requires specific permissions to perform various operations such as creating a conversation, updating a conversation, fetching a user's conversations, deleting a conversation, and adding a message to a conversation.

Also, even if a user has necessary permissions, they cannot perform operations for a different user. Each user can only access their own data.

zmb3 commented on lib/auth/auth_with_roles.go:
```diff
@@ -6423,88 +6423,47 @@ func (a *ServerWithRoles) WatchPendingHeadlessAuthentications(ctx context.Contex
-	}
-
-	return a.authServer.CreateAssistantConversation(ctx, req)
+	return nil, trace.NotImplemented("CreateAssistantConversation must not be called on auth.ServerWithRoles")
```
Can you help me understand why we're breaking from convention and moving RBAC enforcement out of the `with_roles` file?

Joerger replied:
This is the convention shared by the newer, separate gRPC services we've added, such as the trust service. My understanding is that we are trying to move away from a monolithic approach to services.

@rosstimothy @codingllama do you have any more context to add here?

I'm planning on carrying this new approach over to new services that I add in my work for https://github.com/gravitational/teleport/pull/27588, so it'd be good if we agreed on it.

jakule replied:
Because `auth_with_roles` is only being used when the service is implemented as a part of AuthService https://github.com/gravitational/teleport/blob/f0756bdd656b34a9eeb3ad6d6f7618e0de27aa5b/api/proto/teleport/legacy/client/proto/authservice.proto#L2231
New services, such as Okta, DeviceTrust, Assist, etc are implemented as separate GRPC services. In this case, `auth_wit_roles` is not being used, and each service needs to implement the RBAC separately. Here is another example https://github.com/gravitational/teleport/blob/c8fdd9ce93fc7cd93b00c0453a27f6754686ea5f/lib/auth/auth_with_roles.go#L5106
Does it make sense?

zmb3 replied:
Yep, thanks!

zmb3 commented on lib/auth/userpreferences/userpreferencesv1/service.go:
```diff
@@ -21,42 +21,72 @@ import (
+		return nil, trace.Wrap(err)
+	}
+
+	if authCtx.User.GetName() != req.GetUsername() {
```
I thought you just mentioned that there isn't a username in the request?

Joerger replied:
+1 we can remove the username and use `authCtx.User.GetName()` directly

jakule replied:
I mentioned that we addressed the concern by restricting the user to be only able to edit their preferences, but I'll remove the field from the request to avoid further confusion.

jakule replied:
Removed.

Joerger commented on lib/auth/userpreferences/userpreferencesv1/service.go:
```diff
@@ -21,42 +21,72 @@ import (
 
 // UpsertUserPreferences creates or updates user preferences for a given username.
 func (a *Service) UpsertUserPreferences(ctx context.Context, req *userpreferences.UpsertUserPreferencesRequest) (*emptypb.Empty, error) {
+	authCtx, err := authz.AuthorizeWithVerbs(ctx, a.log, a.authorizer, true, types.KindUser, types.VerbUpdate)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if authCtx.User.GetName() != req.GetUsername() {
+		return nil, trace.AccessDenied("user %q is not allowed to access preferences for user %q", authCtx.User.GetName(), req.Username)
+	}
```
I think we want to mimic the `currentUserAction` method used in `auth_with_roles`:

```
// AuthorizeCurrentUser will ensure that the user has access to the given resource based on ownership, or based
// on the given verbs for the given kind.
func AuthorizeCurrentUser(ctx context.Context, log logrus.FieldLogger, authorizer Authorizer, quiet bool, owner string, kind string, verbs ...string) (*Context, error) {
	authCtx, err := authorizer.Authorize(ctx)
	if err != nil {
		return nil, ConvertAuthorizerError(ctx, log, err)
	}

	if authCtx.User.GetName() == owner {
		return authCtx, nil
	}

	ruleCtx := &services.Context{
		User: authCtx.User,
	}

	return authorizeContextWithVerbs(ctx, log, authCtx, quiet, ruleCtx, kind, verbs...)
}
```

```suggestion
	authCtx, err := authz.AuthorizeCurrentUser(ctx, a.log, a.authorizer, true, req.GetUsername(), types.KindUser, types.VerbUpdate)
	if err != nil {
		return nil, trace.Wrap(err)
	}

	if authCtx.User.GetName() != req.GetUsername() {
		return nil, trace.AccessDenied("user %q is not allowed to access preferences for user %q", authCtx.User.GetName(), req.Username)
	}
```

Joerger replied:
Or, given the comment above to remove the username field:

```
func AuthorizeLocalUser(ctx context.Context, log logrus.FieldLogger, authorizer Authorizer, quiet bool) (*Context, error) {
	authCtx, err := authorizer.Authorize(ctx)
	if err != nil {
		return nil, ConvertAuthorizerError(ctx, log, err)
	}

	if _, ok := authCtx.UnmappedIdentity.(LocalUser); !ok {
		return nil, trace.AccessDenied("action not permitted for non-local user")
	}

	return authCtx, nil
}
```

jakule replied:
I think it should be fine now.

Joerger commented on lib/auth/userpreferences/userpreferencesv1/service_test.go:
```diff
@@ -0,0 +1,262 @@
+			},
+			wantErr: assert.Error,
+		},
+		{
+			name:     "access denied - RBAC rule",
+			userName: noAccessUser,
+			req: &userpreferencesv1.UpsertUserPreferencesRequest{
+				Username:    noAccessUser,
+				Preferences: defaultPreferences,
+			},
+			wantErr: assert.Error,
+		},
```
Shouldn't a user be able to update their own preferences?

```suggestion
		{
			name:     "success",
			userName: noAccessUser,
			req: &userpreferencesv1.UpsertUserPreferencesRequest{
				Username:    noAccessUser,
				Preferences: defaultPreferences,
			},
			wantErr: assert.NoError,
		},
```

jakule replied:
In that case, we should probably remove RBAC for user preferences. Just let you do whatever you want with your own preferences. I followed the same rules as the other related resources (password for example).

jakule commented on lib/auth/userpreferences/userpreferencesv1/service_test.go:
```diff
@@ -0,0 +1,262 @@
+
+	roles := map[string]types.Role{}
+
+	role, err := types.NewRole("allow-rules", types.RoleSpecV6{
+		Allow: types.RoleConditions{
+			Rules: []types.Rule{
+				{
+					Resources: []string{types.KindUser},
+					Verbs:     []string{types.VerbList, types.VerbRead, types.VerbUpdate, types.VerbCreate, types.VerbDelete},
+				},
+			},
+		},
+	})
```
It's still a resource. The current implementation was (or at least supposed) to check for User read/update permissions.

Joerger replied:
The current implementation in `auth_with_roles` is using `currentUserAction`, which checks if the user owns the resource OR has User read/update permissions.

This PR changes it so the user must own the resource AND has User read/update permissions, which is not what seems to be intended.

jakule replied:
Changed

jakule committed:
Add missing logger

jakule committed:
"Refactor user preferences request handling"

This commit refactors how GetUserPreferences and UpsertUserPreferences handle requests.
The `username` field is removed from request parameters. Instead of having the client send the user's username in a request, the server now automatically uses the username of the authenticated user making the request.
This change improves the security by preventing a user from attempting to fetch or manipulate another user's preferences. Removed tests were specifically testing the old, insecure behavior.

jakule committed:
Refactor to use authz.HasBuiltinRole

Refactored code in the 'auth_with_roles.go' file to use 'authz.HasBuiltinRole' instead of 'HasBuiltinRole'. This change is in line with recommended practices for deprecation and makes the code more standard and easier to manage. The original 'HasBuiltinRole' function is marked as deprecated and will be removed in future once 'teleport.e' is updated to use 'authz.HasBuiltinRole'.

jakule committed:
Reserve removed username again?

jakule committed:
Fix UT

jakule committed:
Add local user permissions checks in authz

This commit introduces two new methods in permissions.go to check if a user is a local user, and if a given action is performed by a local user. These permission checks are then used to replace existing checks in service.go, when performing actions like creating conversation, updating, listing, etc. This simplifies checks and provides a more consolidated and unified method for verifying user actions.

jakule committed:
Fix tests

jakule committed:
Tweak RBAC

Joerger commented on lib/auth/userpreferences/userpreferencesv1/service.go:
```diff
@@ -21,42 +21,69 @@ import (
 // GetUserPreferences returns the user preferences for a given user.
 func (a *Service) GetUserPreferences(ctx context.Context, req *userpreferences.GetUserPreferencesRequest) (*userpreferences.GetUserPreferencesResponse, error) {
-	return a.backend.GetUserPreferences(ctx, req)
+	authCtx, err := authz.AuthorizeWithVerbs(ctx, a.log, a.authorizer, true, types.KindUser, types.VerbUpdate)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```
We should allow users to get their own user preferences without having `user` `update` privileges.
```suggestion
	authCtx, err := a.authorizer.Authorize(ctx)
	if err != nil {
		return nil, trace.Wrap(err)
	}

	if !authz.IsLocalUser(authCtx) {
		return trace.IsAccessDenied("Non-local user cannot get user preferences")
	}
```

jakule replied:
@zmb3 Any thoughts?

zmb3 replied:
I tend to agree with Brian.

Also why can't SSO users have preferences too?

Joerger replied:
> Also why can't SSO users have preferences too?

`LocalUser` means local cluster user, as opposed to a remote cluster user.

jakule replied:
No objections, changed (with a few tweaks to make it compile :sweat_smile:)

Joerger commented on lib/auth/userpreferences/userpreferencesv1/service.go:
```diff
@@ -21,42 +21,69 @@ import (
 // UpsertUserPreferences creates or updates user preferences for a given username.
 func (a *Service) UpsertUserPreferences(ctx context.Context, req *userpreferences.UpsertUserPreferencesRequest) (*emptypb.Empty, error) {
-	return &emptypb.Empty{}, trace.Wrap(a.backend.UpsertUserPreferences(ctx, req))
+	authCtx, err := authz.AuthorizeWithVerbs(ctx, a.log, a.authorizer, true, types.KindUser, types.VerbUpdate)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```
We should allow users to modify their own user preferences without having `user` `update` privileges.
```suggestion
	authCtx, err := a.authorizer.Authorize(ctx)
	if err != nil {
		return nil, trace.Wrap(err)
	}
	if !authz.IsLocalUser(authCtx) {
		return trace.IsAccessDenied("Non-local user cannot get user preferences")
	}
```

jakule replied:
Changed

Joerger commented on lib/authz/permissions.go:
```diff
@@ -1355,3 +1355,27 @@ func UserFromContext(ctx context.Context) (IdentityGetter, error) {
+	return ok
+}
+
+// IsLocalUserAction checks if the identity is a local user and the username
+func IsLocalUserAction(authContext Context, username string) bool {
+	return IsLocalUser(authContext) && authContext.User.GetName() == username
+}
```
nit: naming
```suggestion
// IsCurrentUser checks if the identity is a local user matching the given username
func IsCurrentUser(authContext Context, username string) bool {
	return IsLocalUser(authContext) && authContext.User.GetName() == username
}
```

jakule replied:
Changed

Joerger commented on lib/auth/userpreferences/userpreferencesv1/service_test.go:
```diff
@@ -0,0 +1,238 @@
+			},
+			wantErr: assert.NoError,
+		},
+		{
+			name:     "access denied - RBAC rule",
+			userName: noAccessUser,
+			req:      &userpreferencesv1.GetUserPreferencesRequest{},
+			want:     nil,
+			wantErr:  assert.Error,
+		},
```
As stated in my comment above, I don't think users should not be denied access to their own user preference resource for not having `user` `update` permissions. If this is left like this, all users will be unable to get/update user preferences, unless they have admin like permissions to update user resources. A user with the `access` role for example would be unable to get/update their user preferences.

Since username is no longer passed separately from the user's certificate for `GetUserPreferences` or `UpsertUserPreferences`, we can expect this method will always succeed for a local user.

If you want, you could instead add a test to ensure that built in role identies are denied access.

jakule committed:
Address review comments

Joerger commented on lib/auth/userpreferences/userpreferencesv1/service_test.go:
```diff
@@ -0,0 +1,232 @@
+
+	roles := map[string]types.Role{}
+
+	role, err := types.NewRole("allow-rules", types.RoleSpecV6{
+		Allow: types.RoleConditions{
+			Rules: []types.Rule{
+				{
+					Resources: []string{types.KindUser},
+					Verbs:     []string{types.VerbList, types.VerbRead, types.VerbUpdate, types.VerbCreate, types.VerbDelete},
+				},
+			},
+		},
+	})
```
It should work without these permissions
```suggestion
	role, err := types.NewRole("allow-rules", types.RoleSpecV6{})
```

jakule replied:
Removed.

Joerger commented on lib/authz/permissions.go:
```diff
@@ -1355,3 +1355,27 @@ func UserFromContext(ctx context.Context) (IdentityGetter, error) {
 	return user, nil
 }
+
+// HasBuiltinRole checks if the identity is a builtin role with the matching
+// name.
+func HasBuiltinRole(authContext Context, name string) bool {
+	if _, ok := authContext.Identity.(BuiltinRole); !ok {
+		return false
+	}
+	if !authContext.Checker.HasRole(name) {
+		return false
+	}
+
+	return true
+}
+
+// IsLocalUser checks if the identity is a local user.
+func IsLocalUser(authContext Context) bool {
+	_, ok := authContext.Identity.(LocalUser)
+	return ok
+}
+
+// IsCurrentUser checks if the identity is a local user matching the given username
+func IsCurrentUser(authContext Context, username string) bool {
+	return IsLocalUser(authContext) && authContext.User.GetName() == username
+}
```
suggestion: Pass by pointer instead of copying it

```suggestion
// HasBuiltinRole checks if the identity is a builtin role with the matching
// name.
func HasBuiltinRole(authContext *Context, name string) bool {
	if _, ok := authContext.Identity.(BuiltinRole); !ok {
		return false
	}
	if !authContext.Checker.HasRole(name) {
		return false
	}

	return true
}

// IsLocalUser checks if the identity is a local user.
func IsLocalUser(authContext *Context) bool {
	_, ok := authContext.Identity.(LocalUser)
	return ok
}

// IsCurrentUser checks if the identity is a local user matching the given username
func IsCurrentUser(authContext *Context, username string) bool {
	return IsLocalUser(authContext) && authContext.User.GetName() == username
}
```

jakule replied:
This function used to take this argument by value, and I can change them to take pointers, I don't think this is the best idea. Pointers can save a few CPU cycles, but then you need to deal with potential `nil` pointer and for loop scope pointer issue.

Joerger commented on lib/auth/userpreferences/userpreferencesv1/service.go:
```diff
@@ -21,42 +21,75 @@ import (
+		return nil, trace.Wrap(err)
+	}
+	if !authz.IsLocalUser(*authCtx) {
+		return nil, trace.AccessDenied("Non-local user cannot get user preferences")
```
nit:
```suggestion
		return nil, trace.AccessDenied("Non-local user cannot upsert user preferences")
```

Joerger committed:
Separate client and server interfaces for user preference services.

Joerger commented on lib/services/userpreferences.go:
```diff
@@ -31,3 +31,11 @@ type UserPreferences interface {
 }
+
+// UserPreferencesBackend is the database interface for managing user preferences.
+type UserPreferencesBackend interface {
```
Suggestion: 9ca8e75bb6f269cc8e89d510403d0635e61920cd

Using the same interface for both the gRPC client and server/backend results in this kind of awkward situation situation. It's better to separate them like most of the new endpoints do.

jakule replied:
Cherry-picked. Thanks

jakule committed:
Apply core review suggestions

Joerger commented on lib/auth/assist/assistv1/service.go:
```diff
@@ -109,33 +127,93 @@ func (a *Service) UpdateAssistantConversationInfo(ctx context.Context, request *
+	}
+
+	// Check if this endpoint is called by a user or Proxy.
+	if _, ok := authCtx.UnmappedIdentity.(authz.LocalUser); ok {
```
```suggestion
	if authz.IsLocalUser(authCtx) {
```

jakule commented on lib/auth/assist/assistv1/service.go:
```diff
@@ -109,33 +127,93 @@ func (a *Service) UpdateAssistantConversationInfo(ctx context.Context, request *
 
 // GetAssistantConversations returns all conversations started by a user.
 func (a *Service) GetAssistantConversations(ctx context.Context, req *assist.GetAssistantConversationsRequest) (*assist.GetAssistantConversationsResponse, error) {
+	authCtx, err := authz.AuthorizeWithVerbs(ctx, a.log, a.authorizer, true, types.KindAssistant, types.VerbRead, types.VerbList)
```
No opinion here. Accepted

jakule committed:
Apply suggestions from code review

Co-authored-by: Brian Joerger <bjoerger@goteleport.com>

jakule commented:
Ping @r0mant @zmb3 @ryanclark

r0mant reviewed (approved):
Bot.
